packages/node/src/indexer/indexer.manager.ts

```diff
@@ -105,9 +108,31 @@ export class IndexerManager extends BaseIndexerManager<
       const { block, events, extrinsics } = blockContent;
       await this.indexBlockContent(block, dataSources, getVM);
 
+      // Group the events so they only need to be iterated over a single time
+      const groupedEvents = events.reduce(
+        (acc, evt, idx) => {
+          if (evt.phase.isInitialization) {
+            acc.init ??= [];
+            acc.init.push(evt);
+          } else if (evt.phase.isFinalization) {
+            acc.finalize ??= [];
+            acc.finalize.push(evt);
+          } else if (evt.extrinsic?.idx) {
+            const idx = evt.extrinsic.idx;
+            acc[idx] = acc[idx] || [];
+            acc[idx].push(evt);
+          } else {
+            logger.warn(
+              `Unrecognized event type, skipping. block="${block.block.header.number.toNumber()}" eventIdx="${idx}"`,
+            );
+          }
+          return acc;
+        },
+        {} as Record<number | 'init' | 'finalize', SubstrateEvent[]>,
```

Comment:
```suggestion
{init: [], finalize: []} as Record<number | 'init' | 'finalize', SubstrateEvent[]>,
```
You can remove other ` ?? []`.